### PR TITLE
Implement login, register, and pull

### PR DESCRIPTION
**Pull request by meganle**
## Overview
Sorry for the semi thick PR I didn't realize I did all of this on just one branch. This PR includes the following features:
- Registering a user which pushes a json to the send-devops repo with their info
- Login involves comparing the given credentials with the user's json on the send-devops repo to verify
- Pulling a config for an app which downloads the docker-compose file for a given app

## Changes Made
- Implemented calls to Github's API
- Auth uses a flow that Github defined for apps which involves generating a JWT token then sending that to receive an installation access token
- The access token and when it expires is written to `~/.send/credentials.json` which is used for subsequent requests
- Also to keep track of whether a user is logged in and which, I write their encrypted username to `~/.send/user`

## Test Coverage
Ran the cli locally and verified login, register, and pull worked

## Next Steps
Start implementing listing apps that a user has access to and granting a user access to an app

### CODE REVIEW

**Review by kevinchan159 — APPROVED** [review]
🔥

File: auth.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	tokenString, err := token.SignedString(signKey)
+	return tokenString
+}
+
+func RequestInstallationToken() string {
+	client := &http.Client{}
+	var bodyBuffer io.Reader = nil
+
+	req, _ := http.NewRequest("POST", "https://github.coecis.cornell.edu/api/v3/app/installations/1/access_tokens", bodyBuffer)
+	req.Header.Set("Authorization", "Bearer "+GenerateJWTToken())
```
Comment (young-k) [inline auth.go]: space around `+`?

File: auth.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+func GetCredentialsPath() string {
+	dir, _ := os.UserHomeDir()
+	return path.Join(dir, ".send")
+}
+
+func WriteCredentials(token string, expires_at int64) {
+	err := os.Mkdir(GetCredentialsPath(), 0755)
+	if err != nil {
+		panic(err)
```
Comment (young-k) [inline auth.go]: wow `panic`, this must be where rust stole this from

File: auth.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	}
+
+	credentials := Credentials{token, expires_at}
+
+	file, _ := json.MarshalIndent(credentials, "", "    ")
```
Comment (young-k) [inline auth.go]: maybe `\t` instead of `""`?

File: git.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+func RegisterUser(username string, password string) {
+	contentsLink := BaseURL + "users/" + username + ".json"
+
+	newUser := User{
+		username,
+		password,
+		[]string{},
+	}
+	userJson, _ := json.MarshalIndent(newUser, "", "	")
```
Comment (young-k) [inline git.go]: would `\t\t` work?

**Review by young-k — APPROVED** [review]
i'm gonna read through go syntax tomorrow, but will approve for now

**Review by young-k — COMMENTED** [review]
@meganle I'm reading through the Go documentation, and it seems that if we don't want to export something from a package we lowercase the first letter of the function/const/type/etc –– do we actually want to export everything from these files? If not, we should probably update the "private" ones to be lowercase.

File: auth.go (hunk 1 above)
Comment (meganle) [reply auth.go]: go's formatter auto formats to this

Comment (young-k) [reply auth.go]: ): why

File: auth.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	if credentials.ExpiresAt < time.Now().Unix() {
+		return requestInstallationToken()
+	}
+
+	return credentials.Token
+}
+
+func WriteUser(username string) {
+	// Encrypt username
+	signBytes := []byte("passphrasewhichneedstobe32bytes!")
```
Comment (young-k) [inline auth.go:97]: sorry im reading thru this again now, what is this for again

File: auth.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+func GetCurrentUser() string {
+	file, err := ioutil.ReadFile(path.Join(getCredentialsPath(), "user"))
+
+	if err != nil {
+		return ""
+	}
+
+	// Decrypt username
+	signBytes := []byte("passphrasewhichneedstobe32bytes!")
```
Comment (young-k) [inline auth.go:115]: should we source this from an `.env` file, or do we have to ensure that it's the same for all builds of the binary?

File: envrc.template [hunk 7]
```diff
@@ -1 +1,3 @@
 export SEND_UPDATES_HOOK_URL=FILL_IN
+export GIT_PEM_KEY_PATH=FILL_IN
+export GIT_APP_ID=FILL_IN
```
Comment (young-k) [inline envrc.template:3]: nit, but this should be alphabetized hahaha

File: git.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	user := user{}
+	json.Unmarshal(fileContents, &user)
+
+	err := bcrypt.CompareHashAndPassword([]byte(user.HashedPassword), password)
+
+	if err != nil {
+		return user, false
+	}
+
+	return user, true
```
Comment (young-k) [inline git.go:133]: could this just be `return user, err != nil`, and we could can drop the last 5 lines?

File: git.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+	req, _ := http.NewRequest(method, url, bodyBuffer)
+	req.Header.Set("Authorization", "token "+GetInstallationToken())
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return nil
+	}
+
+	defer resp.Body.Close()
```
Comment (young-k) [inline git.go:50]: this `defer` syntax is very nifty, i like it a lot

**Review by young-k — COMMENTED** [review]
some minor stuff, i still need to read more but i have a better idea of how this works

File: auth.go (hunk 6 above)
Comment (meganle) [reply auth.go:115]: yeah i was thinking of doing that. we should be able to source from env. will make the fix!

File: auth.go (hunk 5 above)
Comment (meganle) [reply auth.go:97]: it's the key that is used to encrypt the username

File: git.go (hunk 8 above)
Comment (meganle) [reply git.go:133]: yessir

**Pull request merged**